WifiDirect/src/main/java/org/literacy/wifip2p/WifiAccessPoint.java

```diff
@@ -0,0 +1,228 @@
+import java.util.Map;
+
+import static android.net.wifi.p2p.WifiP2pManager.WIFI_P2P_CONNECTION_CHANGED_ACTION;
+
+/**
+ * Created by oscarmakala on 22/07/2016.
+ * Helper for wifi service discovery
+ */
+public class WifiAccessPoint implements
+        WifiP2pManager.ConnectionInfoListener,
+        WifiP2pManager.GroupInfoListener,
+        WifiP2pManager.PeerListListener {
+
+    private static final String AVAILABLE = "available";
+    private static final String TAG = WifiAccessPoint.class.getSimpleName();
+    private static final String SEPERATOR = "|";
+    private final Context mContext;
+    private final WifiP2pManager mWsdManager;
+    private final WifiP2pManager.Channel mChannel;
+
+    public static final String SERVICE_REG_TYPE = "_wdm_p2p._tcp";
+    private WifiP2pManager.ActionListener mActionListener;
+    private BroadcastReceiver mReceiver;
+
+    private String mNetworkName = "";
+    private String mPassphrase = "";
+    private String mInetAddress = "";
+    private IntentFilter intentFilter;
+
+
+    public WifiAccessPoint(Context context) {
+        mContext = context;
+        mWsdManager = (WifiP2pManager) mContext.getSystemService(Context.WIFI_P2P_SERVICE);
+        mChannel = mWsdManager.initialize(mContext, mContext.getMainLooper(), null);
+        mReceiver = new AccessPointReceiver();
+        intentFilter = new IntentFilter();
+        intentFilter.addAction(WifiP2pManager.WIFI_P2P_STATE_CHANGED_ACTION);
+        intentFilter.addAction(WifiP2pManager.WIFI_P2P_CONNECTION_CHANGED_ACTION);
+
+        initialize();
+    }
+
+    /**
+     * If you're providing a local service, you need to register it for service discovery.
+     * Once your local service is registered,
+     * the framework automatically responds to service discovery requests from peers.
+     */
+    public void startLocalService(String instanceName) {
+        Map<String, String> record = new HashMap<>();
+        record.put(AVAILABLE, "visible");
+        WifiP2pDnsSdServiceInfo serviceInfo = WifiP2pDnsSdServiceInfo.newInstance(instanceName, SERVICE_REG_TYPE, record);
+        NotificationCenter.getInstance().postNotificationName(NotificationCenter.LOGGER, TAG, "Add local service :" + instanceName);
+        mWsdManager.addLocalService(mChannel, serviceInfo, mActionListener);
+    }
+
+    public void initialize() {
+        initializeActionListener();
+    }
+
+    private void initializeActionListener() {
+        mActionListener = new WifiP2pManager.ActionListener() {
+            @Override
+            public void onSuccess() {
+                Log.d(TAG, "Added service discovery request");
+            }
+
+            @Override
+            public void onFailure(int code) {
+                if (code == WifiP2pManager.P2P_UNSUPPORTED) {
+                    Log.d(TAG, "P2P isn't supported on this device");
+                }
+            }
+        };
+    }
+
+
+    public void stop() {
+        mContext.unregisterReceiver(mReceiver);
```

Comment:
This lines throws the following exception when the app is closed: 

`java.lang.IllegalArgumentException: Receiver not registered: org.literacy.wifip2p.WifiAccessPoint$AccessPointReceiver@4cac7f7`